cmd/expiration-mailer/main.go

```diff
@@ -229,16 +231,92 @@ func (m *mailer) findExpiringCertificates() error {
 			m.log.Err(fmt.Sprintf("expiration-mailer: Error loading certificates: %s", err))
 			return err // fatal
 		}
-		if len(certs) > 0 {
-			processingStarted := m.clk.Now()
-			m.processCerts(certs)
-			m.stats.TimingDuration("Mailer.Expiration.ProcessingCertificatesLatency", time.Since(processingStarted), 1.0)
+
+		if len(certs) == 0 {
+			m.log.Info(fmt.Sprintf("expiration-mailer: Found 0 certificates that expire between %s and %s and had last nag >%s before expiry", left, right, expiresIn))
+			continue // nothing to do
 		}
+
+		certs = m.eliminateRenewedCerts(certs)
+
+		processingStarted := m.clk.Now()
+		m.processCerts(certs)
+		m.stats.TimingDuration("Mailer.Expiration.ProcessingCertificatesLatency", time.Since(processingStarted), 1.0)
 	}
 
 	return nil
 }
 
+// eliminateRenewedCerts returns a new slice with the already-renewed certificates removed.
+//
+// Structure of goroutines
+//
+//                /-- worker --\
+//  dispatcher ---+-- worker --+--- collector
+//                \-- worker --/
+//
+func (m *mailer) eliminateRenewedCerts(certs []core.Certificate) []core.Certificate {
+	ret := make([]core.Certificate, 0, len(certs))
+	workCh := make(chan core.Certificate)
+	retCh := make(chan core.Certificate)
+	var wg sync.WaitGroup
+
+	query := `
+SELECT 1
+FROM fqdnSets a
+LEFT OUTER JOIN fqdnSets b
+	ON a.setHash = b.setHash
+	AND a.issued < b.issued
+WHERE a.serial = :serial
+`
+	worker := func() {
+		var args = map[string]interface{}{"serial": ""}
+		for cert := range workCh {
+			args["serial"] = cert.Serial
+			var present int64
+			present, err := m.dbMap.SelectInt(query, args)
+			if err != nil {
+				m.log.Err(fmt.Sprintf("expiration-mailer: error fetching renewal state: %v", err))
+				os.Exit(1)
+			} else if present == 0 {
+				retCh <- cert
+			} else {
+				// Drop cert
+			}
+		}
+		wg.Done()
+	}
+
+	// Send work
+	go func() {
+		for _, cert := range certs {
+			workCh <- cert
+		}
+		close(workCh)
+	}()
+
+	// Start workers
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go worker()
+	}
+
+	// Close retCh when done
+	go func() {
+		wg.Wait()
+		close(retCh)
+	}()
+
+	// Collect results
+	for cert := range retCh {
+		ret = append(ret, cert)
+	}
+
+	m.log.Info(fmt.Sprintf("expiration-mailer: dropped %d/%d renewed certificates", len(certs)-len(ret), len(certs)))
```

Comment:
Stating this (even though it's not a super commonly run tool) would be useful as well.